File: test/test.BrowserbaseBrowserLauncher.ts
import { describe, expect, it } from "vitest";

import { BrowserbaseBrowserLauncher } from "../bridge/BrowserbaseBrowserLauncher.js";
import { CdpSocket, expectCdpBrowserSurface } from "./helpers.BrowserLauncher.js";

const LIVE_BROWSERBASE_TIMEOUT_MS = 120_000;
const hasBrowserbaseEnv = Boolean(process.env.BROWSERBASE_API_KEY?.trim());

function browserbaseApiUrl(pathname: string) {
  return new URL(
    pathname,
    `${(process.env.BROWSERBASE_BASE_URL ?? "https://api.browserbase.com").replace(/\/$/, "")}/`,
  );
}

async function retrieveBrowserbaseSession(session_id: string) {
  const response = await fetch(browserbaseApiUrl(`/v1/sessions/${session_id}`), {
    headers: { "x-bb-api-key": process.env.BROWSERBASE_API_KEY! },
  });
  expect(response.status).toBeGreaterThanOrEqual(200);
  expect(response.status).toBeLessThan(300);
  return (await response.json()) as Record<string, unknown>;
}

describe.skipIf(!hasBrowserbaseEnv)("BrowserbaseBrowserLauncher", () => {
  it(
    "creates, verifies, resumes, and releases a real Browserbase browser session",
    { timeout: LIVE_BROWSERBASE_TIMEOUT_MS },
    async () => {
      const launcher = new BrowserbaseBrowserLauncher({
        project_id: process.env.BROWSERBASE_PROJECT_ID,
        timeout: 120,
        ...(process.env.BROWSERBASE_REGION ? { region: process.env.BROWSERBASE_REGION } : {}),
        browser_settings: {
          viewport: { width: 900, height: 700 },
          recordSession: false,
        },
        user_metadata: {
          modcdp_launcher_test: "BrowserbaseBrowserLauncher",
        },
      });
      const browser = await launcher.launch();
      let resumed: Awaited<ReturnType<BrowserbaseBrowserLauncher["launch"]>> | null = null;
      let cdp: CdpSocket | null = null;
      const session_id = browser.browserbase_session_id;

      try {
        expect(session_id).toEqual(expect.any(String));
        expect(browser.browserbase_session_url).toContain(session_id);
        expect(browser.ws_url).toEqual(expect.stringMatching(/^wss:\/\//));
        expect(browser.cdp_url).toBe(browser.ws_url);
        cdp = await CdpSocket.connect(browser.ws_url!);
        await expectCdpBrowserSurface(cdp);

        const retrieved = await retrieveBrowserbaseSession(session_id!);
        expect(retrieved.id).toBe(session_id);
        expect(retrieved.status).toBe("RUNNING");
        if (process.env.BROWSERBASE_PROJECT_ID) expect(retrieved.projectId).toBe(process.env.BROWSERBASE_PROJECT_ID);

        resumed = await new BrowserbaseBrowserLauncher({
          session_id: session_id,
          close_session_on_close: false,
        }).launch();
        expect(resumed.browserbase_session_id).toBe(session_id);
        expect(resumed.ws_url).toEqual(expect.stringMatching(/^wss:\/\//));
        await expectCdpBrowserSurface(cdp);
      } finally {
        await cdp?.close();
        await resumed?.close();
        await browser.close();
        await browser.close();
      }

      await expect
        .poll(async () => (await retrieveBrowserbaseSession(session_id!)).status, { timeout: 30_000, interval: 1_000 })
        .not.toBe("RUNNING");
    },
  );
});

File: python/tests/test_BrowserbaseBrowserLauncher.py
from __future__ import annotations

import json
import os
import time
import unittest
import urllib.request
from typing import Any, cast

from websocket import create_connection

from modcdp.launcher.BrowserbaseBrowserLauncher import BrowserbaseBrowserLauncher


LIVE_BROWSERBASE_TIMEOUT_S = 120


@unittest.skipUnless(os.environ.get("BROWSERBASE_API_KEY", "").strip(), "BROWSERBASE_API_KEY is required for live Browserbase tests")
class BrowserbaseBrowserLauncherTests(unittest.TestCase):
    def test_creates_verifies_resumes_and_releases_real_browserbase_session(self) -> None:
        launcher = BrowserbaseBrowserLauncher(
            cast(Any, {
                "browserbase_project_id": os.environ.get("BROWSERBASE_PROJECT_ID"),
                "timeout": 120,
                **({"region": os.environ["BROWSERBASE_REGION"]} if os.environ.get("BROWSERBASE_REGION") else {}),
                "browserbase_browser_settings": {
                    "viewport": {"width": 900, "height": 700},
                    "recordSession": False,
                },
                "browserbase_user_metadata": {
                    "modcdp_launcher_test": "BrowserbaseBrowserLauncher",
                },
            })
        )
        browser = launcher.launch()
        resumed = None
        ws = None
        session_id = browser.get("browserbase_session_id")
        try:
            if not isinstance(session_id, str):
                self.fail(f"browserbase_session_id = {session_id!r}")
            self.assertIn(session_id, browser.get("browserbase_session_url") or "")
            cdp_url = browser.get("cdp_url")
            if not isinstance(cdp_url, str):
                self.fail(f"cdp_url = {cdp_url!r}")
            self.assertRegex(cdp_url, r"^wss://")
            ws = create_connection(cdp_url, timeout=LIVE_BROWSERBASE_TIMEOUT_S)
            _expect_cdp_browser_surface(ws)

            retrieved = _retrieve_browserbase_session(session_id)
            self.assertEqual(retrieved.get("id"), session_id)
            self.assertEqual(retrieved.get("status"), "RUNNING")
            if os.environ.get("BROWSERBASE_PROJECT_ID"):
                self.assertEqual(retrieved.get("projectId"), os.environ["BROWSERBASE_PROJECT_ID"])

            resumed = BrowserbaseBrowserLauncher(
                {
                    "browserbase_session_id": session_id,
                    "browserbase_close_session_on_close": False,
                }
            ).launch()
            self.assertEqual(resumed.get("browserbase_session_id"), session_id)
            self.assertRegex(resumed.get("cdp_url") or "", r"^wss://")
            _expect_cdp_browser_surface(ws)
        finally:
            if ws is not None:
                ws.close()
            if resumed is not None:
                resumed["close"]()
            browser["close"]()
            browser["close"]()

        deadline = time.time() + 30
        while time.time() < deadline:
            if _retrieve_browserbase_session(session_id).get("status") != "RUNNING":
                return
            time.sleep(1)
        self.fail("Browserbase session did not leave RUNNING status after release")


@unittest.skipIf(os.environ.get("BROWSERBASE_API_KEY", "").strip(), "BROWSERBASE_API_KEY is set")
class BrowserbaseBrowserLauncherWithoutCredentialsTests(unittest.TestCase):
    def test_requires_browserbase_api_key(self) -> None:
        with self.assertRaisesRegex(RuntimeError, "BROWSERBASE_API_KEY"):
            BrowserbaseBrowserLauncher().launch()


def _retrieve_browserbase_session(session_id: str) -> dict:
    request = urllib.request.Request(
        _browserbase_api_url(f"/v1/sessions/{session_id}"),
        headers={"x-bb-api-key": os.environ["BROWSERBASE_API_KEY"]},
    )
    with urllib.request.urlopen(request, timeout=60) as response:
        if response.status < 200 or response.status >= 300:
            raise AssertionError(f"Browserbase session fetch returned {response.status}")
        return json.loads(response.read())


def _browserbase_api_url(pathname: str) -> str:
    base_url = os.environ.get("BROWSERBASE_BASE_URL", "https://api.browserbase.com").rstrip("/")
    return f"{base_url}/{pathname.lstrip('/')}"


def _expect_cdp_browser_surface(ws) -> None:
    ws.send(json.dumps({"id": 1, "method": "Browser.getVersion", "params": {}}))
    message = json.loads(ws.recv())
    if not isinstance(message.get("result", {}).get("product"), str):
        raise AssertionError(f"Browser.getVersion result = {message!r}")


if __name__ == "__main__":
    unittest.main()
